File: JAVA/25_Shortest_Path/BOJ_07_1956_운동.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    final static int INF = Integer.MAX_VALUE / 2;

    public static int floydWarshall(int[][] route, int n) {
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    route[i][j] = Math.min(route[i][j], route[i][k] + route[k][j]);
                }
            }
        }
        int min = INF;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (i != j) {
                    min = Math.min(min, route[i][j] + route[j][i]);
                }
            }
        }
        return min;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        int[][] route = new int[v + 1][v + 1];
        for (int i = 1; i <= v; i++) {
            Arrays.fill(route[i], INF);
            route[i][i] = 0;
        }
        while (e-- > 0) {
            st = new StringTokenizer(br.readLine(), " ");
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            route[a][b] = Math.min(route[a][b], c);
        }
        int min = floydWarshall(route, v);
        System.out.println(min == INF ? -1 : min);
    }
}

File: C/25_Shortest_Path/BOJ_07_1956_운동.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#define math_min(a, b) a < b ? a : b
const int INF = INT_MAX / 2;
int floydWarshall(int** route, int n) {
	for (int k = 1; k <= n; k++) {
		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= n; j++) {
				route[i][j] = math_min(route[i][j], route[i][k] + route[k][j]);
			}
		}
	}
	int min = INF;
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			if (i != j) {
				min = math_min(min, route[i][j] + route[j][i]);
			}
		}
	}
	return min;
}
main() {
	int n, m;
	scanf("%d %d", &n, &m);
	int** route = (int**)malloc(sizeof(int*) * (n + 1));
	for (int i = 1; i <= n; i++) {
		route[i] = (int*)malloc(sizeof(int) * (n + 1));
		for (int j = 1; j <= n; j++) {
			route[i][j] = INF;
		}
		route[i][i] = 0;
	}
	int a, b, c;
	while (m--) {
		scanf("%d %d %d", &a, &b, &c);
		route[a][b] = math_min(route[a][b], c);
	}
	int min = floydWarshall(route, n);
	printf("%d", min != INF ? min : -1);
}